Propagate data export endpoints to webserver

bisgaard-itis opened the pull request:
<!-- Title Annotations:

  WIP: work in progress
  🐛    Fix a bug.
  ✨    Introduce new features.
  🎨    Enhance existing feature.
  ♻️    Refactor code.
  🚑️    Critical hotfix.
  ⚗️    Perform experiments.
  ⬆️    Upgrade dependencies.
  📝    Add or update documentation.
  🔨    Add or update development scripts.
  🔒️    Fix security issues.
  ⚠️    Changes in ops configuration etc. are required before deploying.
        [ Please add a link to the associated ops-issue or PR, such as in https://github.com/ITISFoundation/osparc-ops-environments or https://git.speag.com/oSparc/osparc-infra ]
  🗃️    Database table changed (relevant for devops).
  👽️    Public API changes (meaning: dev features are moved to being exposed in production)
  🚨    Do manual testing when deployed

or from https://gitmoji.dev/
-->

## What do these changes do?

<!-- Badge to openapi specs
[![ReDoc](https://img.shields.io/badge/OpenAPI-ReDoc-85ea2d?logo=openapiinitiative)](https://redocly.github.io/redoc/?url=HERE-URL-TO-RAW-FILE)
-->
- Connects rpc endpoints in storage for exporting data with rest endpoints in webserver
- Introduces "async job" terminology for jobs which are handled by the storage worker. 
- Refactors implementation of async jobs so they will be easier to ruse in the future. 

## Related issue/s

<!-- Link pull request to an issue
  SEE https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

- resolves ITISFoundation/osparc-issues#428
- fixes #26
-->
- closes https://github.com/ITISFoundation/osparc-simcore/issues/7197

## How to test

<!-- Give REVIEWERS some hits or code snippets on how could this be tested -->
```bash
cd services/web
make install dev
cd tests/unit
python -m pytest with_dbs/03/test_storage_rpc.py
```
## Dev-ops checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

<!-- Some checks that might help your code run stable on production, and help devops assess criticality.
Modified from https://oschvr.com/posts/what-id-like-as-sre/

- How can DevOps check the health of the service ?
- How can DevOps safely and gracefully restart the service ?
- How and why would this code fail ?
- What kind of metrics are you exposing ?
- Is there any documentation/design specification for the service ?
- How (e.g. through which loglines) can DevOps detect unexpected situations that require escalation to human ?
- What are the resource limitations (CPU, RAM) expected for this service ?
- Are all relevant variables documented and adjustable via environment variables (i.e. no hardcoded magic numbers) ?
-->

sanderegg commented on api/specs/web-server/Makefile:
```diff
@@ -16,3 +16,6 @@ install-dev install: _check_venv_active
 	python openapi.py
+
+.PHONY: openapi-specs
+openapi-specs: all
```
thought: we should use `uv run` for these kind of stuff.

bisgaard-itis replied:
Yep, I agree.

sanderegg commented on packages/models-library/src/models_library/api_schemas_rpc_async_jobs/async_jobs.py:
```diff
@@ -5,11 +5,11 @@
 from typing_extensions import Self
 
-TaskRpcId: TypeAlias = UUID
+AsyncJobRpcId: TypeAlias = UUID
```
minor: do we need to have Rpc everywhere? it is already in the module for rpc.
also if you want to have that generic, I guess this should be independent of the tech below.

bisgaard-itis replied:
Yes, good point. I will remove the `Rpc` from these.

sanderegg commented on packages/models-library/src/models_library/api_schemas_rpc_async_jobs/async_jobs.py:
```diff
@@ -5,11 +5,11 @@
-class TaskRpcStatus(BaseModel):
-    task_id: TaskRpcId
+class AsyncJobRpcStatus(BaseModel):
+    job_id: AsyncJobRpcId
```
just a side comment, but also think of how we will debug these things.
I'm fine with having a UUID, but that shall be easy to follow-up when I look in a debug log. maybe to keep in mind for later.

bisgaard-itis replied:
Hmm, that's a very good point. One thing we could do is add a prefix/suffix like @pcrespov suggested at some point (I believe inspired by what Stripe does). But from my point of view the right way to do that is really get the tracing up and running properly. When the tracing is enabled the logs are decorated by the `trace_id`, and it should be quite easy to track.

sanderegg commented on packages/models-library/src/models_library/api_schemas_rpc_async_jobs/async_jobs.py:
```diff
@@ -5,11 +5,11 @@
-    task_id: TaskRpcId
+class AsyncJobRpcStatus(BaseModel):
+    job_id: AsyncJobRpcId
     task_progress: PositiveFloat = Field(..., ge=0.0, le=1.0)
```
Can we use `ProgressBarData` object? we have so many flavours. that one I tried to make generic enough so that we can use it everywhere.

GitHK replied:
@sanderegg I guess you mean `ProgressReport` right? why would I want to serialise over the progress bar?

bisgaard-itis replied:
I have corrected this to using a `ProgressReport`

sanderegg commented on packages/models-library/src/models_library/storage_schemas.py:
```diff
@@ -8,16 +8,28 @@
 from typing import Annotated, Any, Literal, Self, TypeAlias
 from uuid import UUID
 
+from models_library.api_schemas_rpc_async_jobs.async_jobs import (
+    AsyncJobRpcGet,
+    AsyncJobRpcResult,
+    AsyncJobRpcStatus,
+)
+from models_library.api_schemas_storage.data_export_async_jobs import (
+    DataExportTaskStartInput,
```
relative imports

bisgaard-itis replied:
Good points. I will fix. Thanks

sanderegg reviewed (approved):
Thanks!

pcrespov reviewed (approved):
thx. Left some suggestions

bisgaard-itis commented on services/storage/src/simcore_service_storage/api/rpc/_async_jobs.py:
```diff
@@ -0,0 +1,41 @@
+@router.expose()
+async def abort(app: FastAPI, job_id: AsyncJobRpcId) -> AsyncJobRpcAbort:
+    assert app  # nosec
+    return AsyncJobRpcAbort(result=True, job_id=job_id)
```
yes

bisgaard-itis commented on packages/models-library/src/models_library/storage_schemas.py:
```diff
@@ -368,3 +380,55 @@ def ensure_consistent_entries(self: Self) -> Self:
+
+
+class AsyncJobGet(BaseModel):
+    job_id: UUID
```
Yep, good point. I will correct.

bisgaard-itis commented on services/web/server/src/simcore_service_webserver/storage/_handlers.py:
```diff
@@ -9,15 +9,22 @@
 from models_library.projects_nodes_io import LocationID
+from models_library.rest_base import RequestParameters
 from models_library.storage_schemas import (
+    AsyncJobGet,
```
I have moved these files into `api_schemas_storage` where there are now a `rest` as well as a `rpc` directory

bisgaard-itis commented on services/web/server/src/simcore_service_webserver/api/v0/openapi.yaml:
done

bisgaard-itis commented on services/web/server/src/simcore_service_webserver/storage/_handlers.py:
```diff
@@ -358,3 +378,101 @@ class _PathParams(BaseModel):
+        ),
+    )
+    return create_data_response(
+        AsyncJobGet.from_async_job_rpc_get(async_job_rpc_get),
```
Very good point. I have adapted this, except that I have not added the `to_domain_model` and `from_domain_model` methods respectively. The reason is that currently there is no real domain models in place here. The webserver essentially just forwards the request to storage directly.

bisgaard-itis commented on services/storage/src/simcore_service_storage/api/rpc/_async_jobs.py:
```diff
@@ -0,0 +1,41 @@
+)
+from servicelib.rabbitmq import RPCRouter
+
+router = RPCRouter()
```
Very good point. I will add this.

bisgaard-itis replied:
Done

bisgaard-itis commented on services/web/server/src/simcore_service_webserver/storage/_handlers.py:
```diff
@@ -358,3 +377,100 @@ class _PathParams(BaseModel):
     return create_data_response(payload, status=resp_status)
+
+
+@routes.post(
```
Yep, I added an endpoint in storage and propagated it to webserver as well

GitHK commented on api/specs/web-server/_storage.py:
```diff
@@ -167,3 +174,49 @@ async def is_completed_upload_file(
+    """Get async job status"""
+
+
+@router.get(
+    "/storage/async-jobs",
+    response_model=Envelope[list[StorageAsyncJobGet]],
+    name="get_async_job_result",
+)
+async def get_async_jobs():
+    """Get async job status"""
```
The UserID is also missing below.
```suggestion
@router.get(
    "/storage/async-jobs",
    response_model=Envelope[list[StorageAsyncJobGet]],
    name="get_async_jobs",
)
async def get_async_jobs(user_id: UserID):
    """Retrunsa list of async jobs for the user"""
```

bisgaard-itis replied:
Thanks

GitHK commented on api/specs/web-server/_storage.py:
```diff
@@ -167,3 +174,49 @@ async def is_completed_upload_file(
+    name="get_async_job_result",
+)
+async def get_async_job_result(storage_async_job_get: StorageAsyncJobGet, job_id: UUID):
+    """Get async job status"""
```
```suggestion
    """Get the result of the async job"""
```

GitHK commented on api/specs/web-server/_storage.py:
```diff
@@ -167,3 +174,49 @@ async def is_completed_upload_file(
+@router.post(
+    "/storage/locations/{location_id}/export-data",
+    response_model=Envelope[StorageAsyncJobGet],
+    name="storage_export_data",
```
```suggestion
    name="export_data",
```

GitHK commented on services/storage/src/simcore_service_storage/api/rpc/_data_export.py:
```diff
@@ -1,55 +1,38 @@
-    assert task_id  # nosec
-    return TaskRpcResult(result="Here's your result.", error=None)
+    assert user_id  # nosec
+    return [AsyncJobGet(job_id=uuid4(), job_name="myjob")]
```
I will then change all of the names in a final PR to be the same

bisgaard-itis replied:
But I would prefer to keep it a uuid. What do you mean there are 3 different types? The `job_id` is uuid everywhere. I might have a typealias to make the code cleaner to read, but that is also a uuid under the hood. I would keep it a uuid. I guess the point is it has to be unique, so I would prefer to have a type which enforces this to the extend that it is possible. I could see an idea in using a format <uuid>_<human readable prefix to indicate it is an async job>, but then I would add a check against a regex pattern to check that this is fulfilled.

GitHK reviewed (commented):
The listing and start of a job could also be made generic.
I think that this would be ideal in our situation.

This way we don't need to write each time a different RPC interface. We just provide a task name and its parameters.

It if cannot be started an error will be raised by the worker (which is fine).

bisgaard-itis commented:
> The listing and start of a job could also be made generic. I think that this would be ideal in our situation.
> 
> This way we don't need to write each time a different RPC interface. We just provide a task name and its parameters.
> 
> It if cannot be started an error will be raised by the worker (which is fine).

That's a good point. I have moved the `list_jobs` endpoint to the generic implementation of the async jobs. Then I introduced filtering á la Docker (https://docs.docker.com/reference/api/engine/version/v1.47/#tag/Container/operation/ContainerList) which should be generic enough that we can have a single endpoint for different types of tasks. Concerning the endpoint for triggering a task, I am hesitant to move it to a generic implementation. The reasons are the exception types. I think on the client side (meaning the webserver here) it is useful that exceptions specific to what kind of error is going on can be propagated. That is quite specific to different types of tasks. So if OK with you I will leave it as is for now. We can always move later if you prefer.

There is one challenge which is still not entirely clear to me how to handle: Permission and access rights. From talking with Sylvain I think this should be handled in storage, but for that we might need to propagate more info from the webserver (e.g. `project_id`s)

GitHK commented on packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/storage/data_export.py:
```diff
@@ -23,51 +17,12 @@
-
-
-async def get_data_export_result(
-    rabbitmq_rpc_client: RabbitMQRPCClient, *, task_id: TaskRpcId
-) -> TaskRpcResult:
-    result = await rabbitmq_rpc_client.request(
-        STORAGE_RPC_NAMESPACE,
-        _RPC_METHOD_NAME_ADAPTER.validate_python("get_data_export_result"),
-        task_id=task_id,
-        timeout_s=_DEFAULT_TIMEOUT_S,
-    )
-    assert isinstance(result, TaskRpcResult)
+    assert isinstance(result, AsyncJobGet)
```
I would do something like this and make it reusable. 
We achive two things:
- we can easily resue it for let's say: copy of folder a folder, copy of a project, etc...
- we can more easily extract all this logic to a common library in the future

```suggestion
from datetime import timedelta
AsyncJobName: TypeAlias = str
async def starsubmit_job(
    rabbitmq_rpc_client: RabbitMQRPCClient, *, name: AsyncJobName, timeout: timedelta, **kwargs
) -> AsyncJobGet:
    result = await rabbitmq_rpc_client.request(
        STORAGE_RPC_NAMESPACE,
        TypeAdapter(RPCMethodName).validate_python(name),
        paths=paths,
        timeout_s=timeout.total_seconds(),
    )
    assert isinstance(result, AsyncJobGet)
```

GitHK replied:
So even thouhg on the server part the `start_data_export` still needs to be defined as it currently is, on the client side we use this method and call it like so. This would be in the http API on the webserver.

```python
await submit_job(rabbit_rpc_client, name="start_data_export", paths=data_export_task_start_input)
```

bisgaard-itis replied:
I see your point, but TBH I would prefer not to do this, if you agree. This function can reach any rpc endpoint exposed within the platform. I find it more intuitive that a single client function can hit a single server endpoint. In my opinion we should use something like openapi specs in the future to check compatibility between clients and servers also for rpc. And that would not work with this setup.

bisgaard-itis replied:
We do not require openapi specs no, but we should use it (https://spec.open-rpc.org/). To me the specs are like type hinting in python, they allow you to automate checks that apis and clients are compatible or that apis don't break backwards compatibility. So I think we should use it also for the rpc. But that doesn't work if things are too dynamic. But ok if you are adamant to get have a general method I will add it.

bisgaard-itis replied:
Done

GitHK reviewed (commented):
one more thing and it should be all set

matusdrobuliak66 commented on services/web/server/src/simcore_service_webserver/storage/_rest.py:
```diff
@@ -358,3 +382,142 @@ class _PathParams(BaseModel):
+@permission_required("storage.files.*")
+@handle_data_export_exceptions
+async def get_async_job_status(request: web.Request) -> web.Response:
+    class _RequestContext(RequestParameters):
```
Q: Why are you defining the class always in each route? why not to have in the top that you reuse in the module?

bisgaard-itis replied:
I factored this out now. Thanks

matusdrobuliak66 commented on services/web/server/src/simcore_service_webserver/storage/_rest.py:
```diff
@@ -358,3 +382,142 @@ class _PathParams(BaseModel):
+@handle_data_export_exceptions
+async def abort_async_job(request: web.Request) -> web.Response:
+    class _RequestContext(RequestParameters):
+        user_id: UserID = Field(..., alias=RQT_USERID_KEY)  # type: ignore[literal-required]
```
Q: The product concept is not important on these routes? Are those jobs not bound to a product?

bisgaard-itis replied:
Probably it is 😅 . I change the code so it is forwarded to storage. Thanks

GitHK reviewed (commented):
thanks a lot!
